sycl/include/CL/sycl/handler.hpp

```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
 #endif
   }
 
+  /// Implements a command group function that enqueues a kernel that calls
+  /// user's lambda function \param KernelFunc and does one iteration of
+  /// reduction of elements in each of work-groups.
+  /// This version uses tree-reduction algorithm to reduce elements in each
+  /// of work-groups. At the end of each work-group the partial sum is written
+  /// to a global buffer.
+  ///
+  /// Briefly: user's lambda, tree-reduction, CUSTOM types/ops.
+  template <typename KernelName, typename KernelType, int Dims, class Reduction>
+  void reduCGFunc(KernelType KernelFunc, const nd_range<Dims> &Range,
+                  Reduction &Redu) {
+
+    size_t NWorkItems = Range.get_global_range().size();
+    size_t WGSize = Range.get_local_range().size();
+    size_t NWorkGroups = Range.get_group_range().size();
+
+    bool IsUnderLoaded = (NWorkGroups * WGSize - NWorkItems) != 0;
+    bool IsEfficientCase = !IsUnderLoaded && ((WGSize & (WGSize - 1)) == 0);
+
+    bool IsUpdateOfUserAcc =
+        Reduction::accessor_mode == access::mode::read_write &&
+        NWorkGroups == 1;
+
+    // Use local memory to reduce elements in work-groups into 0-th element.
+    // If WGSize is not power of two, then WGSize+1 elements are allocated.
+    // The additional last element is used to catch elements that could
+    // otherwise be lost in the tree-reduction algorithm.
+    size_t NumLocalElements = WGSize + (IsEfficientCase ? 0 : 1);
+    auto LocalReds = Redu.getReadWriteLocalAcc(NumLocalElements, *this);
```

Comment:
That is one of reasons why I have that func in reduction_impl class.
Because the buffer.hpp and accessor.hpp cannot be included in handler.hpp.
I'll change if you really insist/repeat the request, I would prefer this 'auto'.